### PR TITLE
Remove `RamAccountingContext`, use new eagerly ram accounting instead

### PR DESCRIPTION
**Pull request by seut**
The old `RamAccountingContext` uses a flush buffer to prevent flushing to
the circuit breaker on every accounted bytes. This leads to a possible
underaccounting.
Additionally the implementation isn't thread-safe but was used
from multiple threads which also leads to wrong accounting.

This implementation is superseeded by new `RamAccounting` implementations
which are either thread-safe (`ConcurrentRamAccounting`) or eagerly allocate
(on-demand) a block of bytes (`BlockBasedRamAccounting`, this one must be
created dedicated per thread).

The last commit adds a test assertion to ensure that the `Query` circuit breaker is decreased on every job/statement additionally to existing breaker assertions.

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/fetch/TransportFetchOperation.java [hunk 1]
```diff
@@ -64,8 +68,22 @@ public TransportFetchOperation(TransportFetchNodeAction transportFetchNodeAction
             nodeId,
             nodeIdToReaderIdToStreamers.get(nodeId),
             new NodeFetchRequest(jobId, fetchPhaseId, closeContext, toFetch),
-            ramAccounting,
+            ramAccountingForIncomingResponse(ramAccounting, toFetch, closeContext),
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/fetch/TransportFetchOperation.java]: We could also decide if we should account inside the `NodeFetchResponse` only `if (numReaders > 0)` https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/execution/engine/fetch/NodeFetchResponse.java#L56.

**Comment by seut** [comment]
Somehow GH does not like my re-ordering of commits, the 3rd commit must be 2nd, otherwise the 2nd commit results in flaky tests. Locally the commits are in the right order 🤔

File: sql/src/main/java/io/crate/execution/jobs/JobSetup.java [hunk 2]
```diff
@@ -278,9 +279,16 @@ private void registerContextPhases(Iterable<? extends NodeOperation> nodeOperati
             }
             if (ExecutionPhases.hasDirectResponseDownstream(nodeOperation.downstreamNodes())) {
                 var executionPhase = nodeOperation.executionPhase();
-                var ramAccounting = RamAccountingContext.forExecutionPhase(breaker(), executionPhase);
+                CircuitBreaker breaker = breaker();
+                int ramAccountingBlockSizeInBytes = BlockBasedRamAccounting.calculateBlockSizeInBytes(
+                    breaker.getLimit(),
+                    1
+                );
+                var ramAccounting = ConcurrentRamAccounting.forCircuitBreaker(executionPhase.label(), breaker);
```
[the remark this answers is not in the document]
Comment (seut) [reply sql/src/main/java/io/crate/execution/jobs/JobSetup.java]: right, good catch

File: sql/src/main/java/io/crate/execution/jobs/JobSetup.java [hunk 3]
```diff
@@ -886,33 +924,33 @@ public Boolean visitNestedLoopPhase(NestedLoopPhase phase, Context context) {
...
+                breaker.getLimit(),
+                1
+            );
+            var ramAccounting = ConcurrentRamAccounting.forCircuitBreaker(phase.label(), breaker);
+            var ramAccountingOfOperation = new BlockBasedRamAccounting(
+                ramAccounting::addBytes,
+                ramAccountingBlockSizeInBytes);
+            RowConsumer lastConsumer = context.getRowConsumer(phase, Paging.PAGE_SIZE, ramAccountingOfOperation);
+            var memoryManager = memoryManagerFactory.getMemoryManager(ramAccounting);
+            var memoryManagerLeft = memoryManagerFactory.getMemoryManager(ramAccounting);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/jobs/JobSetup.java]: Same here regarding the MemoryManager instances. Sorry I had only pointed it out once as I wasn't sure if I'm missing something.

Comment (seut) [reply sql/src/main/java/io/crate/execution/jobs/JobSetup.java]: ah thanks!

**Pull request merged**